### PR TITLE
chore: Switch to stable Rust

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion-comet/issues/142

## Rationale for this change

Continues the work started by @sunchao in https://github.com/apache/datafusion-comet/pull/373

### CODE REVIEW

File: core/src/parquet/util/hash_util.rs [hunk 1]
```diff
@@ -135,6 +135,7 @@ mod tests {
     use super::*;
 
     #[test]
+    #[ignore] // thread caused non-unwinding panic. aborting.
```
Comment (andygrove) [inline core/src/parquet/util/hash_util.rs:138]:
```
unsafe precondition(s) violated: slice::from_raw_parts requires the pointer to be aligned and non-null, and the total size of the slice not to exceed `isize::MAX`
```

File: rust-toolchain [hunk 2]
```diff
@@ -1 +1 @@
-nightly-2023-09-05
+1.77
```
Comment (andygrove) [inline rust-toolchain:1]: I will file a separate issue for upgrading to 1.78. We will need to fix some assumptions in our unsafe code first.

Comment (andygrove) [reply rust-toolchain:1]: I filed https://github.com/apache/datafusion-comet/issues/507

Comment (kazuyukitanimura) [reply rust-toolchain:1]: Will this use `1.77.2` automatically? Looks like there are few fixes

**Review by kazuyukitanimura — APPROVED** [review]
LGTM

File: rust-toolchain (hunk 2 above)
Comment (andygrove) [reply rust-toolchain]: I updated this to 1.77.2

File: core/src/common/mutable_vector.rs [hunk 3]
```diff
@@ -1,409 +0,0 @@
...
-use crate::{
-    common::{bit, CometBuffer, ValueGetter, ValueSetter},
-    BinaryType, StringType, StringView, TypeTrait,
-};
-
-const DEFAULT_ARRAY_LEN: usize = 4;
-
-/// A mutable vector that can be re-used across batches.
-#[derive(Debug)]
-pub struct MutableVector {
```
Comment (viirya) [inline core/src/common/mutable_vector.rs:31]: We don't need this?

**Review by viirya — APPROVED** [review]

File: core/src/common/mutable_vector.rs (hunk 3 above)
Comment (andygrove) [reply core/src/common/mutable_vector.rs:31]: @sunchao Removed this in the original PR that I forked from.

Comment (sunchao) [reply core/src/common/mutable_vector.rs:31]: Yes we don't need this anymore. It was created originally to facilitate Comet specific data types, vectors and kernels (for instance leveraging array view), but we've decided to fully use DataFusion now.

**Review by sunchao — APPROVED** [review]
LGTM, thanks @andygrove for taking this over!

**Comment by andygrove** [comment]
> LGTM. Thanks for taking this over. I was about to ask chao if he is still working on the #373.
> 
> 
> 
> It looks like the CI is still failed, you may take a look at that:
> 
> ```
> 
> - unary negative integer overflow test *** FAILED *** (817 milliseconds)
> 
>   "[ARITHMETIC_OVERFLOW] short overflow. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error. SQLSTATE: 22003" did not contain " caused overflow" (CometExpressionSuite.scala:1563)
> 
> ```

It looks like the unary minus tests are failing in main as well for Spark 4. I think there was a conflict between two PRs that were merged yesterday.

**Pull request merged**